PennMobile/src/main/java/com/pennapps/labs/pennmobile/viewmodels/LaundryViewModel.kt

```diff
@@ -0,0 +1,250 @@
+                        hallName = "Lauder"
+                    }
+
+                    // add the hall name to the list
+                    laundryHalls.add(hallName)
+                    laundryRooms[hallName] = roomList
+                }
+            } else {
+                Log.i("Laundry", "Failed to get laundry rooms")
+            }
+
+            _loadedRooms.postValue(true)
+        }
+    }
+    fun existsDiff() : Boolean {
+        var diff = false
+        runBlocking {
+            favoritesMutex.withLock {
+                if (_favoriteRooms.value!!.favoriteRooms.size != curToggled.size) {
+                    diff = true
+                    return@runBlocking
+                }
+                for (room in _favoriteRooms.value!!.favoriteRooms) {
+                    if (!curToggled.contains(room.id)) {
+                        diff = true
+                        return@runBlocking
+                    }
+                }
+            }
+        }
+        return diff
+    }
+
+    private suspend fun sendPreferences(studentLife: StudentLife_rf2, bearerToken: String,
+                                        favoriteIdList: List<Int>) {
+        val laundryRequest = LaundryRequest(favoriteIdList)
+        val response = studentLife.sendLaundryPref(bearerToken, laundryRequest)
+        if (response.isSuccessful) {
+            Log.i("Laundry Preferences", "Successfully updated preferences")
+        } else {
+            Log.i("Laundry Preferences", "Error updating preferences")
+        }
+    }
+    fun setFavoritesFromToggled(studentLife: StudentLife_rf2, bearerToken: String) {
+        CoroutineScope(Dispatchers.IO).launch {
+            // make a copy of the set
+            val favoriteIdList = curToggled.toList()
+            populateFavorites(coroutineContext, studentLife, favoriteIdList)
+
+            // send
+            sendPreferences(studentLife, bearerToken, favoriteIdList)
+        }
+    }
+
+    fun setToggled() {
+        curToggled.clear()
+        runBlocking {
+            favoritesMutex.withLock {
+                for (room in _favoriteRooms.value!!.favoriteRooms) {
+                    curToggled.add(room.id)
+                }
+            }
+        }
+    }
+
+    // returns true if there is a change state
+    fun toggle(roomId: Int) : Boolean {
+        val origState = curToggled.size >= maxNumRooms
+        if (curToggled.contains(roomId)) {
+            curToggled.remove(roomId)
+
+        } else {
+            curToggled.add(roomId)
+        }
+        return (curToggled.size >= maxNumRooms).xor(origState)
+    }
+    fun getGroupCount() : Int {
+        return laundryHalls.size
+    }
+    fun getChildrenCount(i: Int): Int {
+        return laundryRooms[laundryHalls[i]]!!.size
+    }
+    fun getGroup(i: Int): Any {
+        return laundryHalls[i]
+    }
+    fun getChild(i: Int, i1: Int): Any {
+        return laundryRooms[laundryHalls[i]]!![i1]
```

Comment:
check null

Reply:
this feels janky but prob null checks will help

Reply:
I think it's actually ok to crash here.